File: src/linked_lists/java/ReturnKthToLast.java
package linked_lists.java;

import static linked_lists.java.Node.printLinkedList;
import static linked_lists.java.RemoveDups.initLinkedList;

import java.util.Scanner;

/**
 *
 */
public class ReturnKthToLast {

  public static void main(String[] args) {
    Node list = initLinkedList(10);
    System.out.println("Linked list:");
    printLinkedList(list);

    System.out.println("Input the number:");
    Scanner in = new Scanner(System.in);
    int     n  = in.nextInt();
    System.out.printf("The element is %d.", kthToLast(list, n).data);
  }

  private static Node kthToLast(Node head, int k) {
    if (head == null) {
      throw new IllegalArgumentException("Null linked list.");
    }

    Node fast = head;
    for (int i = 0; i < k; i++, fast = fast.next) {
      if (fast == null) {
        throw new IllegalArgumentException("List too short for index " + k);
      }
    }

    Node slow = head;
    while (fast != null) {
      slow = slow.next;
      fast = fast.next;
    }

    return slow;
  }
}

File: src/linked_lists/kotlin/ReturnKthToLast.kt
package linked_lists.kotlin

/**
 *
 */
fun main(args: Array<String>) {
    val list = initLinkedList(10)
    println("Linked list:")
    printLinkedList(list)

    println("Input the number:")
    val k = readLine().toString().toInt()
    println("The element is " + kthToLast(list, k).data)
}

fun kthToLast(head: Node, k: Int): Node {
    var fast = head as? Node
    for (i in 1..k) {
        if (fast == null) throw IllegalArgumentException("List is too short.")
        fast = fast.next
    }
    var slow = head as? Node
    while (fast != null) {
        slow = slow!!.next
        fast = fast.next
    }

    return slow!!
}